WT-15856 Rework layered cursors positioning

ivankochin opened the pull request:
There is an issue with the current positioning logic for `next()` and `prev()`. According to the comment in `__clayered_iterate_constituent`, it expects either none or both of the constituents to be positioned.

However, there are cases where other functions that are supposed to leave the cursor positioned end up returning only one cursor positioned (which is also set as `current_cursor`). For example, this could happen if `search()` finds the entry in the ingest table, so it doesn’t need to access the stable table in this case. As a result, `stable_cursor` is left unpositioned. This can lead to a situation where the cursor could return out-of-order keys. Please refer to the information in the ticket for more details.

This PR reworks the positioning logic to make it more straightforward and better documented. It also adds documentation to establish a clear contract between different interfaces regarding leaving a cursor positioned.

Additionally, it introduces extra testing for layered cursors, which we previously did not have, and moves the testing oplog implementation to `helper_disagg.py` to allow sharing across multiple tests.

ivankochin committed:
WT-15856 Init commit

ivankochin committed:
Working version + extended test

ivankochin committed:
First clean-up

ivankochin committed:
Revert changes in test_layered23

ivankochin committed:
Add docs

ivankochin committed:
Pass checks

jiechenbo commented on src/cursor/cur_layered.c:
```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
-        else
-            alternate_cursor = clayered->stable_cursor;
+    do {
+        WT_ERR(__clayered_iterate_constituents(clayered, iter_flag, deleted));
```
Would like some comments around this loop to understand what it's trying to achieve

ivankochin replied:
I added a comment and a FIXME as it was discussed

korteland commented on src/cursor/cur_layered.c:
```diff
@@ -609,16 +609,13 @@ __clayered_open_cursors(WT_SESSION_IMPL *session, WT_CURSOR_LAYERED *clayered, b
 static int
-__clayered_get_current(
-  WT_SESSION_IMPL *session, WT_CURSOR_LAYERED *clayered, bool smallest, bool *deletedp)
+__clayered_get_current(WT_SESSION_IMPL *session, WT_CURSOR_LAYERED *clayered, bool smallest)
```
This function has changed for the better -- thanks. It always bothered me a bit that this did a bunch more than it said on the tin.

korteland commented on src/cursor/cur_layered.c:
```diff
@@ -936,20 +904,31 @@ __layered_prev_int(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
 static int
 __layered_prev(WT_CURSOR *cursor)
```
I really like what you did with `__layered_prev` and `__layered_next` -- they're much easier to reason about now.

korteland reviewed (approved):
Wow -- this is a great change. LGTM!

ivankochin committed:
Early return when one of constituents doesn't exist + position alternate basing on the current

Early return is important for the case when the case when a constituent hasn't been created
on the first operation, but then before the next, the creation occured so we still might need
to position the alternate cursor.

Positioning constituent should be done based on the current cursor position and not on the
iface cursor key position for the same reason.

ivankochin committed:
Add WT_RET for __clayered_position_alternate

ivankochin committed:
Handle WT_NOTFOUND for alternate cursor

jiechenbo commented on src/cursor/cur_layered.c:
```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
+    WT_CURSOR *c_stable = clayered->stable_cursor;
+
+    /* At least one cursor is expected to be initialized. */
+    WT_ASSERT(session, c_stable != NULL || c_ingest != NULL);
```
I'm pretty sure we always expect both c_stable and c_ingest to be initalized. Was there a case where this is not the case?

ivankochin replied:
AFAIK one of the constituents cursors could be NULL when the corresponding table doesn't exist yet (for example on the follower when the stable table hasn't been created yet)

ivankochin replied:
I added a FIXME comment there as discussed

jiechenbo commented on src/cursor/cur_layered.c:
```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
+     * WT_CURSTD_KEY_INT being set on the user-visible cursor is the main indication that the cursor
+     * is positioned.
+     */
+    if (!F_ISSET(&clayered->iface, WT_CURSTD_KEY_INT) && !deleted) {
```
Is it possible for the iface cursor is set to WT_CURSTD_KEY_EXT but it is deleted? That seems strange to me

ivankochin replied:
The `!deleted` part protects us from entering the unpositioned flow (which simply calls next() on both constituents) in cases where the previous iteration (the outer while loop) detected a deleted key.

This protection is needed, for example, in the following scenario:
```
cursor = table.open_cursor()  # Opens an unpositioned cursor
cursor.next()
    # The first __clayered_iterate_constituents() call enters this `if` block
    # because neither WT_CURSTD_KEY_INT nor WT_CURSTD_KEY_EXT is set for iface,
    # so it calls next() on both constituents.
    #
    # Then __clayered_deleted() returns deleted == true, indicating that the
    # current key has been deleted.
    #
    # We then call __clayered_iterate_constituents() a second time to advance
    # to the next key. At this point, the iface key is still unset, but we no
    # longer want to enter this `if` block because both constituent cursors are
    # already positioned.
```

jiechenbo commented on src/cursor/cur_layered.c:
```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
+     * The cursor is positioned, but `iter_flag` is not set so we cannot rely on alternate cursor
+     * and need to position it.
+     */
+    if (!F_ISSET(clayered, iter_flag)) {
```
style: Don't need braces around one-line if statement.

ivankochin replied:
Done

jiechenbo commented on src/cursor/cur_layered.c:
```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
+    }
+
+    /* If the alternate cursor's key is equal to the current one, we should move it as well. */
+    if (F_ISSET(c_alternate, WT_CURSTD_KEY_INT)) {
```
Trying to construct a scenario with above. So if we have keys (3, 5) and only stable cursor was positioned.

If the stable cursor is positioned at 3, the ingest cursor should be positioned at 5 at this point? And then at L820, we move stable cursor to 5?

ivankochin replied:
Discussed IRL

jiechenbo commented on src/cursor/cur_layered.c:
```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
+    WT_RET_NOTFOUND_OK(__clayered_constituent_iter(c_current, forward));
 
+done:
+    if (!F_ISSET(clayered, iter_flag)) {
```
What does this if statement do? If it was previously not set to WT_CLAYERED_ITERATE_PREV or WT_CLAYERED_ITERATE_NEXT, shouldn't we set it now bvecause we fixed iterating next/prev?

ivankochin replied:
Discussed IRL

ivankochin committed:
Add FIXMEs

jiechenbo reviewed (approved):
LGTM!
